### PR TITLE
Dqp

### PR DESCRIPTION
**Pull request by Debarati-Pandit**

### CODE REVIEW

File: measure/src/main/scala/org/apache/griffin/measure/utils/TableDetails.scala [hunk 1]
```diff
@@ -32,4 +32,15 @@ object TableDetails {
...
     saveMode
   }
+  
+  def getPartitions(dbName: String, tableName: String): DataFrame ={
+     val df = {
+      if(dbName.trim.nonEmpty)
+        spark.sql(s"show partitions $dbName.$tableName")
+      else spark.sql(s"show partitions $tableName")
+    }
+    df
```
Comment (tanvirp) [inline measure/src/main/scala/org/apache/griffin/measure/utils/TableDetails.scala]: Do you need to declare `df`? Shouldn't just the `if - else` be enough?

File: service/src/main/resources/application.properties [hunk 2]
```diff
@@ -68,7 +68,7 @@ livy.uri=http://192.168.9.12:8998/batches
 yarn.uri=http://192.168.9.12:8088
 
 hdp.kerberos.enabled=false
-hdp.kerberos.user-keytab=/home/enquero/keytabs/hdpqqe.user.keytab
+hdp.kerberos.userKeytab=/home/enquero/keytabs/hdpqqe.user.keytab
```
Comment (tanvirp) [inline service/src/main/resources/application.properties]: Please revert this change? `ServiceConfig` is reading the property with `-`.

File: measure/src/main/scala/org/apache/griffin/measure/step/builder/DataFrameOpsDQStepBuilder.scala [hunk 3]
```diff
@@ -29,11 +29,15 @@ case class DataFrameOpsDQStepBuilder() extends RuleParamStepBuilder {
   def buildSteps(context: DQContext, ruleParam: RuleParam): Seq[DQStep] = {
     val name = getStepName(ruleParam.getOutDfName())
     val inputDfName = getStepName(ruleParam.getInDfName())
-    val transformStep = if (AnomalyType.desc.equalsIgnoreCase(ruleParam.getDqType.desc)) DataFrameOpsTransformStep(
+    /*val transformStep = if (AnomalyType.desc.equalsIgnoreCase(ruleParam.getDqType.desc)) DataFrameOpsTransformStep(
```
[the remark this answers is not in the document]
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/step/builder/DataFrameOpsDQStepBuilder.scala]: Removed the commented code

File: measure/src/main/scala/org/apache/griffin/measure/step/builder/DataFrameOpsDQStepBuilder.scala [hunk 4]
```diff
@@ -29,11 +29,15 @@ case class DataFrameOpsDQStepBuilder() extends RuleParamStepBuilder {
...
       name, inputDfName, ruleParam.getRule, ruleParam.getDetails, ruleParam.getCache, context.name, context.dataSources, context.sinkParams)
     else DataFrameOpsTransformStep(
       name, inputDfName, ruleParam.getRule, ruleParam.getDetails, ruleParam.getCache)
-    transformStep +: buildDirectWriteSteps(ruleParam)
+    transformStep +: buildDirectWriteSteps(ruleParam)*/
+    if (AnomalyType.desc.equalsIgnoreCase(ruleParam.getDqType.desc)) Seq(DataFrameOpsTransformStep(
+      name, inputDfName, ruleParam.getRule, ruleParam.getDetails, ruleParam.getCache, context.name, context.dataSources, context.sinkParams))
+    else DataFrameOpsTransformStep(
+      name, inputDfName, ruleParam.getRule, ruleParam.getDetails, ruleParam.getCache) +: buildDirectWriteSteps(ruleParam)
```
[the remark this answers is not in the document]
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/step/builder/DataFrameOpsDQStepBuilder.scala]: It is supposed to work in the else part. For code readability, have put braces in else.

File: measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetection.scala [hunk 5]
```diff
@@ -16,7 +16,7 @@ import org.apache.griffin.measure.configuration.enums.SourceType.SourceType
 */
```
[the remark this answers is not in the document]
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetection.scala]: As already mentioned in the comment, the object initiates the anomaly detection computation.

File: measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala [hunk 6]
```diff
@@ -40,10 +43,38 @@ object AnomalyDectectionOps {
     
     println(s"Job Name:${jobName}")
```
[the remark this answers is not in the document]
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala]: Have extended Loggable trait in the required objects and replaced println() with info().

File: measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala [hunk 7]
```diff
@@ -40,10 +43,38 @@ object AnomalyDectectionOps {
...
+    println("where = "+where)
+    
+    var allPartitionsProcessed = false
+    val whereExpr = {
+      if(where.isEmpty){
+         val whereExprArr = Utilities.getWhereExprBasedOnPrevPartitionRuns(jobName, dbName, tableName)
+         println("SR whereExprArr => "+whereExprArr.isEmpty)
+         println("SR whereExprArr String => "+whereExprArr.mkString(","))
+         //If whereExprArr is empty, then all the partitions are already run, so stop
+         if(whereExprArr.isEmpty) allPartitionsProcessed = true
```
[the remark this answers is not in the document]
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala]: Made changes as suggested

File: measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala [hunk 8]
```diff
@@ -80,10 +119,13 @@ object AnomalyDectectionOps {
...
-      aggregateExprList, filterRowLevelExpr, filterAggrLevelExpr, medianColumn, outputSourceType,
+    AnomalyDetection.initAnomalyDetection(spark, jobName, inputSourceType, dbName, tableName, groupByList,
+      aggregateExprList, filterExprWithPartition, filterAggrLevelExpr, medianColumn, outputSourceType,
       outputDBName, outputTableName, considerLowerLimit, considerUpperLimit, computationTypeValue)
-    //}
+      if(where.isEmpty){
+        Utilities.savePartitionsProcessed(jobName, whereExpr)
+      }
+    }
     return null
```
[the remark this answers is not in the document]
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala]: Made changes as suggested

File: measure/src/main/scala/org/apache/griffin/measure/utils/TableDetails.scala (hunk 1 above)
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/utils/TableDetails.scala]: Removed the declaration of df

File: service/src/main/resources/application.properties (hunk 2 above)
Comment (Debarati-Pandit) [reply service/src/main/resources/application.properties]: I could still see ServiceConfig is reading it without -. Also while starting the application, getting error if I keep the - in the properties file.

File: measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala [hunk 9]
```diff
@@ -40,10 +43,38 @@ object AnomalyDectectionOps {
...
+         println("SR whereExprArr String => "+whereExprArr.mkString(","))
+         //If whereExprArr is empty, then all the partitions are already run, so stop
+         if(whereExprArr.isEmpty) allPartitionsProcessed = true
+         whereExprArr.mkString(",")
+      }else where
+    }
     
-    val intermediateTableName = sourceConnector.getConfig("table.name").toString()
+    println("SR allPartitionsProcessed => "+allPartitionsProcessed)
+    if(!allPartitionsProcessed){
```
[the remark this answers is not in the document]
Comment (Debarati-Pandit) [reply measure/src/main/scala/org/apache/griffin/measure/step/transform/AnomalyDetectionOps.scala]: Have to check how to refactor this code. The anomaly detection algorithm is executed only when this if condition is true.

File: service/src/main/resources/application.properties (hunk 2 above)
Comment (tanvirp) [reply service/src/main/resources/application.properties]: What error are you getting?

Comment (Debarati-Pandit) [reply service/src/main/resources/application.properties]: Caused by: java.lang.IllegalArgumentException: Could not resolve placeholder 'hdp.kerberos.userKeytab' in string value "${hdp.kerberos.userKeytab}"
	at org.springframework.util.PropertyPlaceholderHelper.parseStringValue(PropertyPlaceholderHelper.java:174) ~[spring-core-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.util.PropertyPlaceholderHelper.replacePlaceholders(PropertyPlaceholderHelper.java:126) ~[spring-core-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.core.env.AbstractPropertyResolver.doResolvePlaceholders(AbstractPropertyResolver.java:236) ~[spring-core-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.core.env.AbstractPropertyResolver.resolveRequiredPlaceholders(AbstractPropertyResolver.java:210) ~[spring-core-4.3.6.RELEASE.jar:4.3.6.RELEASE]
	at org.springframework.context.support.PropertySourcesPlaceholderConfigurer$2.resolveStringValue(PropertySourcesPlaceholderConfigurer.java:172) ~[spring-context-4.3.6.RELEASE.jar:4.3.6.RELEASE]

Comment (tanvirp) [reply service/src/main/resources/application.properties]: I have taken care or this issue. Take the latest pull and try again.
